src/Compiler/Service/FSharpCheckerResults.fs

```diff
@@ -2357,12 +2357,25 @@ module internal ParseAndCheckFile =
         let tokenizer =
             LexFilter.LexFilter(indentationSyntaxStatus, options.CompilingFSharpCore, Lexer.token lexargs true, lexbuf, false)
 
-        (fun _ -> tokenizer.GetToken())
+        if ct.CanBeCanceled then
+            (fun _ ->
+                ct.ThrowIfCancellationRequested()
```

Comment:
How big of an impact on perf does checking for cancellation at every token have? That seems _very_ granular - should we instead check every N tokens/end of line/etc?

Reply:
I was thinking about this as well, but in the end this is what the check does:
https://referencesource.microsoft.com/#mscorlib/system/threading/CancellationTokenSource.cs,716cb764064ecfeb

And I would need to replace it with mutable counter management (increase, compare) and doing e.g. a modulo N operation.

Reply:
Maybe `let` and `member`?

Reply:
in .fsi files both can be fully missing I guess?

Reply:
The cancellation check is really a simple op, you might actually force me for proving that :-D

Reply:
`member` would be in fsi file as well, so I suppose also add `val` ? But maybe that's unnecessarily granular, so perhaps `type`?

Reply:
Now it definitely feels overcomplicated just to balance a simple token check.
Both 'type' as well as 'member' can be fully missing in a file.

Reply:
Would it be cheap to check a token line here? We could probably check if cancellation needed when line changes, or even every few lines.

Reply:
I did a loop from 0 to 10_000, which is only doing a basic multiplication by two in every operation - a single instruction, so this should really highlight the cost differences between cancellation token check or anything more complex.

NaiveCheckNoneToken  - check each iteration, but it is a CancellationToken.None
NaiveCheckRealToken  - check each iteration, but this token can be cancelled
CheckFewHardcodedOps  - this does basic branching similar to how this would need to branch on type of token
CheckRealTokenEveryFewOps  - this does need to keep a separate counter, and do a 'modulo` calculation on every iteration 

|                    Method | every |      Mean |     Error |    StdDev | Allocated |
|-------------------------- |------ |----------:|----------:|----------:|----------:|
|       NaiveCheckNoneToken |     ? |  6.216 us | 0.1164 us | 0.1032 us |         - |
|       NaiveCheckRealToken |     ? |  7.581 us | 0.0774 us | 0.0604 us |         - |
|      CheckFewHardcodedOps |     ? |  6.635 us | 0.0793 us | 0.0662 us |         - |
| CheckRealTokenEveryFewOps |    10 | 13.275 us | 0.0790 us | 0.0700 us |         - |
| CheckRealTokenEveryFewOps |   100 | 13.718 us | 0.1547 us | 0.1372 us |         - |
| CheckRealTokenEveryFewOps |   500 | 13.076 us | 0.2246 us | 0.1991 us |         - |
| CheckRealTokenEveryFewOps |  1000 | 12.932 us | 0.1433 us | 0.1340 us |         - |



=> `ct.ThrowIfCancellationRequested()` is really as lightweight as it can get, I do not think we should bother about it at all.

Reply:
NOTE: This was really a 1-op situation repeated 10_000 times, hence this really shows just the overhead differences in a very artificial scenario.
Even though real lexing is super fast, there are still a few more things going on (deciding based on characters, indexing into tables, allocating a token type,...) - the ThrowIfCancellationRequested() really should not be a worry in the case of not being cancelled.


Things do get a cost hit when the operation is being cancelled, that is given - all of this PR is under the assumption that things in the IDE toolchain would deal with cancellation later anyway, so the cost for throwing exception will always be exist somewhere, and with this PR it will just happen sooner.

Reply:
Thanks for investigating this - I'm now convinced that the naieve approach is best :)